loader/src/reference_binder.py

```diff
@@ -0,0 +1,150 @@
+import re
+from collections import defaultdict
+from dotty_dict import dotty
+
+from loader.src.config.logger import create_logger
+
+logger = create_logger("reference_binder")
+
+
+# dotty-dict does not handle brackets indices,
+# it uses dots instead (a.0.b instead of a[0].b)
+def dotty_paths(paths):
+    for path in paths:
+        yield re.sub(r"\[(\d+)\]", r".\1", path)
+
+
+class ReferenceBinder:
+    def __init__(self, fhirstore):
+        self.fhirstore = fhirstore
+
+        # cache is a dict of form
+        # {
+        #   (fhir_type_target, value, system): {(fhir_type_source, path, isArray): [fhir_id1, ...]},
+        #   (fhir_type_target, value, system): {(fhir_type_source, path, isArray): [fhir_id]},
+        #   ...
+        # }
+        # eg:
+        # {
+        #   (Practitioner, 1234, system): {(Patient, generalPractitioner, True): [fhir-pract-id]},
+        #   ...
+        # }
+
+        self.cache = defaultdict(lambda: defaultdict(list))
+
+    def resolve_references(self, unresolved_fhir_object, reference_paths):
+        fhir_object = dotty(unresolved_fhir_object)
+
+        # iterate over the instance's references and try to resolve them
+        for reference_path in dotty_paths(reference_paths):
+            logger.debug(
+                f"Trying to resolve reference for resource {fhir_object['id']} "
+                f"at {reference_path}"
+            )
+            try:
+                bound_ref = self.bind_existing_reference(fhir_object, reference_path)
+                fhir_object[reference_path] = bound_ref
+            except Exception as e:
+                logger.warning(
+                    "Error while binding reference for instance "
+                    f"{fhir_object} at path {reference_path}: {e}"
+                )
+
+        if fhir_object.get("identifier") and len(fhir_object["identifier"]) > 0:
```

Comment:
`if "identifier" in fhir_object` is clearer

Reply:
done